### PR TITLE
Fixed #26819 -- Fixed crash in ModelForm.validate_unique() on complex types like list.

### PR DESCRIPTION
**Pull request by PREM1980**
https://code.djangoproject.com/ticket/26819

### CODE REVIEW

File: tests/postgres_tests/migrations/0002_create_test_models.py [hunk 1]
```diff
@@ -182,6 +182,27 @@ class Migration(migrations.Migration):
...
+        ),
+        migrations.CreateModel(
+            name='MapSpot',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('position', ArrayField(base_field=models.IntegerField(), size=2)),
+                ('map', models.ForeignKey(on_delete=models.CASCADE, to='postgres_tests.Map')),
+            ],
+            options={
+                'indexes': [],
```
Comment (timgraham) [inline tests/postgres_tests/migrations/0002_create_test_models.py]: You need to add `'required_db_vendor': 'postgresql',` in `options` so the model with `ArrayField` isn't created on other databases.

**Comment by timgraham** [comment]
The tests is failing on PostgreSQL too because it looks like you forgot to commit the fix.

**Comment by timgraham** [comment]
Since the test isn't PostgreSQL specific, it would be nice if the test could live outside `postgres_tests`, perhaps in `model_forms`, by writing a custom field that normalizes to a list.

**Comment by PREM1980** [comment]
Little bit confused on the custom field that normalizes to a list. This issue happens only for inlineformsets.
So, I thought I will add a fix in the model_informsets and here is the working fix that Im planning to add.

def test_inlineformset_unique_validation_with_arrayfield(self):
        call_command('migrate', 'postgres_tests', verbosity=0)
        mapspotformset = inlineformset_factory(Map, MapSpot, can_delete=False, extra=3, fields="__all__")
        data = {
            'mapspot_set-TOTAL_FORMS': '3',  # the number of forms rendered
            'mapspot_set-INITIAL_FORMS': '0',  # the number of forms with initial data
            'mapspot_set-MAX_NUM_FORMS': '',  # the max number of forms
            'mapspot_set-0-position': '10,20',
            'mapspot_set-1-position': '10,20',
            'mapspot_set-2-position': '30,40',
        }
        map_obj = Map.objects.create(name='SanFrancisco')
        formset = mapspotformset(data, instance=map_obj)
        self.assertEqual(formset.errors, [{}, {'__all__': ['Please correct the duplicate values below.']}, {}])

**Comment by PREM1980** [comment]
I'm looking at the custom fields that you had mentioned in the docs..Will get back to you on that.


One more question, the tests under model_formsets does not have any migrations. How do I add my new migrations?

**Comment by timgraham** [comment]
You don't need to add migrations for the models in apps without migrations. The models will be created automatically.

**Comment by PREM1980** [comment]
Tim,
THis issue is more specific to postgres. It happens only in ArrayField.

I can define a default field that gets converted as list but the problem would be to save the data.

**Comment by timgraham** [comment]
This gives an idea of how to implement the field I have in mind: http://stackoverflow.com/a/22343764/5112. It needs some updating since `SubfieldBase` is now removed. You can store the data in a text field by casting the list to text.

**Comment by PREM1980** [comment]
Tim, 
I tried over-riding the to_python and get_db_prep_value method to convert the input string(for example:- 10,20) to list but it does not work for the unique_check constraint.

When the unique check constraint is executed it checks the pre-processed value(that will be entered in the database) which is a string. 

In my case, I need that to be a list.Please let me know if I need to add this test very specific to postgres for now.

Thanks.

**Comment by timgraham** [comment]
Could you push the (broken) code that you wrote so far to this branch so I can check it?

**Comment by PREM1980** [comment]
Here is my changes and its pushed.

https://github.com/PREM1980/django/commit/e109f84a9c06adfa62d1d387359548d99c7d88bf

Here is the sequence of events to refresh -

1. The changes that I implemented in django/forms/models.py is working for
postgres db specific tests.

2. You mentioned to write a generic test that will be applicable to all the
database not just specific to postgres.

3. Thats where I have the problem.

4. I created a field that will take a string input value of
'10,20'(example:- mapspot) and convert that into a list - [10,20] for
processing as a list. Again, I reconvert this list to a string to save into
the database.

When the code checks for unique constraint it always uses the value that is
saved into the database ie; it uses the string.

To test my fix for all database, the list value should be used.

I have given the models.py(tests/model_formsets/generic_test_model.py) alone
that I created to test this condition and not the whole project.

Hope the information that I have given helps.

Let me know for any additional details.


Kind regards,

Prem
408-393-2545



On Tue, Aug 23, 2016 at 5:33 PM, Tim Graham <notifications@github.com>
wrote:
Could you push the (broken) code that you wrote so far to this branch so I
can check it?

**Comment by timgraham** [comment]
Thanks for the example. So what I think you actually need is a custom form field (rather than model field) that normalizes to a list (as `django.django.contrib.postgres.forms.SimpleArrayField` does). You can then write a custom `ModelForm` that uses that field and pass the form class to `inlineformset_factory (parent_model, model, form=...)`.

**Comment by timgraham** [comment]
Closing due to inactivity.

**Pull request closed without merging**